activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -131,7 +131,12 @@ def primary_key(klass)
         end
 
         def foreign_key_present?
-          owner._read_attribute(reflection.foreign_key)
+          if reflection.foreign_key.is_a?(Array)
+            # composite foreign keys branch
+            reflection.foreign_key.all? { |fk| owner._read_attribute(fk) }
+          else
+            owner._read_attribute(reflection.foreign_key)
+          end
```

Comment:
Why not do linear code here?
```suggestion
          Array(reflection.foreign_key).all? { |fk| owner._read_attribute(fk) }
```

Reply:
Most likely this is how I prototyped it when it wasn't clear what is going to stay and a branch was giving a clearer picture of what is necessary. Eventually `reflection.foreign_key` will be an array so neither `is_a` neither `Array()` won't be necessary. Though while it is here perhaps we should prefer the branch as it's faster not including unnecessary allocations from `Array()` 
It's a micro benchmark but for backwards compatibility code which is ought to be removed perhaps we should stick to a slightly more efficient option unless it hurts readability. 

```ruby
require "benchmark/ips"

def if_branch(value)
  if value.is_a?(Array)
    value.all? { |v| v.odd? }
  else
    value.odd?
  end
end

def array_wrap(value)
  Array(value).all? { |v| v.odd? }
end

VALUES = [1, 2, [3, 4]]

Benchmark.ips do |x|
  x.report("Array()") { VALUES.each { |v| array_wrap(v) } }

  x.report("if branch") { VALUES.each { |v| if_branch(v) } }

  x.compare!
end
```

Reply:
I am not a committer, but I disagree that explicit `is_a?` calls are better for readability. If anything, it makes the code look more cluttered and decreases understandability. Especially if your goal is to eliminate the checks, eliminating `Array()` calls is much less error-prone than eliminating branching correctly.

As for performance, I doubt that you can argue the performance both ways, on one side saying that `is_a?` checks don't matter since the value is memoized anyway, and then presenting a benchmark that is doing the same thing over and over again, with no memoization.

This might be a more representative benchmark:
```ruby
require "benchmark/ips"

def if_branch(value)
  @if_branch = nil if Random.rand < 0.10
  @if_branch ||= if value.is_a?(Array)
    value.all? { |v| v.odd? }
  else
    value.odd?
  end
end

def array_wrap(value)
  @array_wrap = nil if Random.rand < 0.10
  @array_wrap ||= Array(value).all? { |v| v.odd? }
end

VALUES = [1, 2, [3, 4]]

Benchmark.ips do |x|
  x.report("Array()") { VALUES.each { |v| array_wrap(v) } }

  x.report("if branch") { VALUES.each { |v| if_branch(v) } }

  x.compare!
end
```
where the memoized value should be hit ~90% of the time, and it gives a slight slowdown, at best:
```
Warming up --------------------------------------
             Array()   228.277k i/100ms
           if branch   258.324k i/100ms
Calculating -------------------------------------
             Array()      2.317M (± 2.0%) i/s -     11.642M in   5.027179s
           if branch      2.557M (± 2.7%) i/s -     12.916M in   5.055496s

Comparison:
           if branch:  2556859.9 i/s
             Array():  2316859.7 i/s - 1.10x  (± 0.00) slower
```

Reply:
Sorry It's a little confusing as this PR contains several different methods that include `if` or `Array()` backwards-compatibility logic and I agree that it's fair to include memoization in the benchmark for cases like `active_record_primary_key` or `foreign_key` though `join_id_for` and `foreign_key_present?` methods do not fall into memoization flow and are being called more frequently. 
I think there is a better argument to make in favor of `Array()` - method signatures. Even though we don't have explicit method signatures here, `if` approach results in the method signature like (sorbet-syntax) `T.any(String, T::Array[String])` which is generally harder to maintain.

So we end up turning `if` branches into `Array()` where possible. For `foreign_key_present?` it didn't matter as the method returns boolean regardless but for `join_id_for` it will be `T::Array[String]` only

Reply:
@paracycle I've reviewed this with Nikita this morning and we got rid of some  branching, but not all of them are removable right now because in some places Active Record expects a single ID not an array. The next steps will be to change the callers to expect an array so we can remove the branching in these methods. I think this PR is good for now (we're trying to keep changes small and contained). I understand the next steps so I'm confident we'll get to a place where we can remove the branching.